index.js

```diff
@@ -155,179 +132,110 @@ function readProtocolData (reader) {
-  return { requestId, protocolData }
-}
-
-function deserializeClpPacket (buffer) {
-  const {typeString, packet} = _toTypeString(buffer)
-  return {
-    type: buffer[0],
-    typeString,
-    packet
+  switch (obj.type) {
+  case TYPE_ACK:
+  case TYPE_RESPONSE:
+  case TYPE_MESSAGE:
+    writeProtocolData(writer, obj.data.protocolData)
+  break
+
+  case TYPE_ERROR: {
+    const ilpPacket = maybeSerializeIlpError(obj.data.rejectionReason)
+    writer.write(ilpPacket)
+    writeProtocolData(writer, obj.data.protocolData)
+  } break
+
+  case TYPE_PREPARE: {
+    const transferIdBuffer = Buffer.from(obj.data.transferId.replace(/\-/g, ''), 'hex')
+    const amountAsPair = stringToTwoNumbers(obj.data.amount)
+    const executionConditionBuffer = Buffer.from(obj.data.executionCondition, 'base64')
+    const expiresAtBuffer = toGeneralizedTimeBuffer(obj.data.expiresAt)
+    writer.write(transferIdBuffer)
+    writer.writeUInt64(amountAsPair)
+    writer.write(executionConditionBuffer)
+    writer.writeVarOctetString(expiresAtBuffer)
+    writeProtocolData(writer, obj.data.protocolData)
+  } break
+
+  case TYPE_FULFILL: {
+    const transferIdBuffer = Buffer.from(obj.data.transferId.replace(/\-/g, ''), 'hex')
+    const fulfillmentBuffer = Buffer.from(obj.data.fulfillment, 'base64')
+    writer.write(transferIdBuffer)
+    writer.write(fulfillmentBuffer)
+    writeProtocolData(writer, obj.data.protocolData)
+  } break
+
+  case TYPE_REJECT: {
+    const transferIdBuffer = Buffer.from(obj.data.transferId.replace(/\-/g, ''), 'hex')
+    const rejectionReasonBuffer = maybeSerializeIlpError(obj.data.rejectionReason)
+    writer.write(transferIdBuffer)
+    writer.write(rejectionReasonBuffer)
+    writeProtocolData(writer, obj.data.protocolData)
+  } break
+
+  default:
+    throw new Error('Unrecognized type')
   }
+
+  const envelopeWriter = new Writer()
+  envelopeWriter.writeUInt8(obj.type)
+  envelopeWriter.writeUInt32(obj.requestId)
+  envelopeWriter.writeVarOctetString(writer.getBuffer())
+  return envelopeWriter.getBuffer()
 }
 
-function _toTypeString (buffer) {
-  switch (buffer[0]) {
-    case TYPE_ACK: return {
-      typeString: 'clp_ack',
-      packet: deserializeAck(buffer) }
-    case TYPE_RESPONSE: return {
-      typeString: 'clp_response',
-      packet: deserializeResponse(buffer) }
-    case TYPE_ERROR: return {
-      typeString: 'clp_error',
-      packet: deserializeError(buffer) }
-    case TYPE_PREPARE: return {
-      typeString: 'clp_prepare',
-      packet: deserializePrepare(buffer) }
-    case TYPE_FULFILL: return {
-      typeString: 'clp_fulfill',
-      packet: deserializeFulfill(buffer) }
-    case TYPE_REJECT: return {
-      typeString: 'clp_reject',
-      packet: deserializeReject(buffer) }
-    case TYPE_MESSAGE: return {
-      typeString: 'clp_message',
-      packet: deserializeMessage(buffer) }
-    default:
-      throw new Error('Packet has invalid type')
+function deserialize (buffer) {
```

Comment:
I am not a fan of collapsing everything in one method. It reduces readability.

Also, you removed `typeString`.

Reply:
OK, I'll reinstate the separate methods.

About the typeString, what is it needed for? It just duplicates the type constant, right?
Adding that as a field just breaks it away from the asn.1 format, so that's quite ugly. What is the use case, just to make the logs more readable? Would it be ok if I just add a static function `typeToString` to the module exports, that converts from e.g. `ClpPacket.TYPE_ACK` to the string 'ack'?

Reply:
Yes, `typeString` is handy for logging. Replacing it with a function is fine.

Reply:
However, the current return type (i.e. containing a `typeString`) is consistent with `ilp-packet`. Although I agree that a static function would be cleaner, we might want to keep `typeString` for consistency. (or change `ilp-packet` also to a static function if that does not break too many dependencies.)

Reply:
ok. i'll open an issue on ilp-packet because i do like it a lot better as a separate function. This way, the object returned by deserialize is the same as the object you need to give to serialize. Adding `typeString` inside the return object would mean `obj === deserialize(serialize(obj))` would no longer be true.